File: Python/BOJ/2504.py
arr = list(input())

isError = False
tmp = 1; result = 0
stack = []
for i in range(len(arr)):
  char = arr[i]
  if char == "(" or char == "[":
    if char == "(": tmp *= 2
    else: tmp *= 3
      
    stack.append(char)
  else:
    if not stack:
      isError = True
      break

    if char == ")" and stack[-1] == "(":
      stack.pop()
      if arr[i-1] == "(":
        result += tmp
      tmp /= 2
    elif char == "]" and stack[-1] == "[":
      stack.pop()
      if arr[i-1] == "[":
        result += tmp
      tmp /= 3
    else:
      isError = True
      break
  
if isError or stack:
  print("0")
else:
  print(int(result))

File: Swift/BOJ/문제집/스택/2504.swift
//
//  2504.swift
//  Algorithm
//
//  Created by 유정주 on 2023/07/08.
//

import Foundation

let arr = Array(readLine()!)

var dict: [Character: Int] = ["(": 2, "[": 3]
var tmp = 1, result = 0
var stack: [Character] = []
var isError = false
for i in 0..<arr.count {
    let char = arr[i]
    if char == "(" || char == "[" {
        tmp *= dict[char]!
        stack.append(char)
    } else {
        guard let top = stack.last else {
            isError = true
            break
        }
        
        if (char == ")" && top == "(") || (char == "]" && top == "[") {
            stack.removeLast()
            if arr[i-1] == top {
                result += tmp
            }
            tmp /= dict[top]!
        } else {
            isError = true
            break
        }
    }
}

if !stack.isEmpty || isError {
    print("0")
} else {
    print(result)
}
